### PR TITLE
Adding Firebase

### PR DESCRIPTION
**Pull request by ethangj**

### CODE REVIEW

File: _general/integrations/firebase.md [hunk 1]
```diff
@@ -0,0 +1,107 @@
+---
+title: Integrating Codeship With Firebase
+shortTitle: Integrating Codeship With Firebase
+tags:
+  - integrations
+  - db
+  - databases
+  - fire base
```
Comment (joesiewert) [inline _general/integrations/firebase.md]: `fire base` > `firebase`?

File: _general/integrations/firebase.md [hunk 2]
```diff
@@ -0,0 +1,107 @@
...
+    title: Using Firebase
+    weight: 19
+---
+
+* include a table of contents
+{:toc}
+
+## About Firebase
+
+[Firebase](https://firebase.google.com/) is a cloud platform for handing operations and data for mobile and web applications.
```
Comment (joesiewert) [inline _general/integrations/firebase.md]: `handing` > `handling`

File: _general/integrations/firebase.md [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+
+The [Firebase documentation](https://firebase.google.com/docs/) provides a great guide to getting started, and the instructions below have more information on integrating with [Codeship Basic](https://codeship.com/features/basic) and [Codeship Pro](https://codeship.com/features/pro).
+
+## Codeship Pro
+
+### Installing Firebase
+
+To use Firebase with Codeship Pro, you'll need to install the `firebase-tools` CLI.
+
+This CLI is packaged as an NPM, so you will need a service defined in your [codeship-services.yml file]({{ site.baseurl }}{% link _pro/builds-and-configuration/services.md %}) with Node and NPM installed.
```
Comment (joesiewert) [inline _general/integrations/firebase.md]: `as an NPM` > `with NPM`

File: _general/integrations/firebase.md [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+
+Inside the service with Node and NPM installed, you will next need to either add `firebase-tools` to your `package.json` and run `npm install` inside your Dockerfile, or add the following line directly to your Dockerfile:
+
+```bash
+RUN npm install firebase --save
```
Comment (joesiewert) [inline _general/integrations/firebase.md]: `firebase` > `firebase-tools`?

File: _general/integrations/firebase.md [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+After installing the Firebase Tools CLI into your service, you will need to add at least your `FIREBASE_TOKEN` via the [encrypted environment variables]({{ site.baseurl }}{% link _pro/builds-and-configuration/environment-variables.md %}) that you encrypt and include in your [codeship-services.yml file]({{ site.baseurl }}{% link _pro/builds-and-configuration/services.md %}).
+
+Depending on how your app uses Firebase, or how you deploy to Firebase, you may want to set other environment variables as well. Some values you may need to set are:
+
+- `apiKey`
+- `authDomain`
+- `databaseURL`
+- `storageBucket`
+
+What environment variables are required depends on which Firebase services and commands you are trying to you, and the [Firebase documentation](https://firebase.google.com/docs/) can provide more specific instructions.
```
Comment (joesiewert) [inline _general/integrations/firebase.md]: `trying to you` > `trying to use`

File: _general/integrations/firebase.md [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+You can do this by navigating to _Project Settings_ and then clicking on the _Environment_ tab.
+
+Depending on how your app uses Firebase, or how you deploy to Firebase, you may want to set other environment variables as well. Some values you may need to set are:
+
+- `apiKey`
+- `authDomain`
+- `databaseURL`
+- `storageBucket`
+
+What environment variables are required depends on which Firebase services and commands you are trying to you, and the [Firebase documentation](https://firebase.google.com/docs/) can provide more specific instructions.
```
Comment (joesiewert) [inline _general/integrations/firebase.md]: `trying to you` > `trying to use`

File: _general/integrations/firebase.md [hunk 7]
```diff
@@ -0,0 +1,107 @@
...
+
+### Using Firebase
+
+If your main application uses Firebase as a database or for other services, simply installing the Firebase Tools CLI and setting the required environment variables via your [environment variables]({{ site.baseurl }}{% link _basic/builds-and-configuration/set-environment-variables.md %}) should be all that is required.
+
+However, if you need to pass the Firebase Tools CLI specific commands - such as deploying your static site to Firebase - you can simply add those commands to either your [setup, test or deployment commands]({{ site.baseurl }}{% link _basic/quickstart/getting-started.md %}) .
```
Comment (joesiewert) [inline _general/integrations/firebase.md]: Extra space > ` .`

File: _general/integrations/firebase.md (hunk 1 above)
Comment (ethangj) [reply _general/integrations/firebase.md]: That on is there for people who search for it incorrectly.

Comment (mlocher) [reply _general/integrations/firebase.md]: Could we add a tag for `firebase` as well?

File: _general/integrations/firebase.md [hunk 8]
```diff
@@ -0,0 +1,107 @@
...
+### Installing Firebase
+
+To use Firebase with Codeship Pro, you'll need to install the `firebase-tools` CLI.
+
+This CLI is packaged with NPM, so you will need a service defined in your [codeship-services.yml file]({{ site.baseurl }}{% link _pro/builds-and-configuration/services.md %}) with Node and NPM installed.
+
+Inside the service with Node and NPM installed, you will next need to either add `firebase-tools` to your `package.json` and run `npm install` inside your Dockerfile, or add the following line directly to your Dockerfile:
+
+```bash
+RUN npm install firebase-tools --save
```
Comment (mlocher) [inline _general/integrations/firebase.md]: We don't need the `--save` here. It is used for modifying the `package.json` file and adding the dependency to that file, which isn't something we want to do during a CI run.

File: _general/integrations/firebase.md [hunk 9]
```diff
@@ -0,0 +1,107 @@
...
+If your main application uses Firebase as a database or for other services, simply installing the Firebase Tools CLI in the service the application runs in and setting the required environment variables via your [encrypted environment variables]({{ site.baseurl }}{% link _pro/builds-and-configuration/environment-variables.md %}) should be all that is required.
+
+However, if you need to pass the Firebase Tools CLI specific commands - such as deploying your static site to Firebase - you can simply add those commands as steps in your [codeship-steps.yml file]({{ site.baseurl }}{% link _pro/builds-and-configuration/steps.md %}).
+
+For example:
+
+```yaml
+- name: Deploy To Firebase
+  service: app
+  command: firebase deploy --token "$FIREBASE_TOKEN" --project $PROJECT_NAME
```
Comment (mlocher) [inline _general/integrations/firebase.md]: that command doesn't work as is, as variable substitution is only available if the command is run within a shell. We'd either need to extract that into a script or run it via a shell, i.e.

```
command: bash -c "firebase deploy --token ${FIREBASE_TOKEN} --project ${PROJECT_NAME}"
```

(I removed the quotes around the variables, as I'm not sure how to escape them, and they aren't required, but really good practice.)

File: _general/integrations/firebase.md [hunk 10]
```diff
@@ -0,0 +1,107 @@
...
+
+**Note** that in this example `app` would be the service defined in the [codeship-services.yml file]({{ site.baseurl }}{% link _pro/builds-and-configuration/services.md %}) with the Firebase Tools CLI installed.
+
+## Codeship Basic
+
+### Installing Firebase
+
+To use Firebase with Codeship Basic, you'll need to install the `firebase-tools` CLI using `npm` in your project's [setup commands]({{ site.baseurl }}{% link _basic/quickstart/getting-started.md %}).
+
+**Note** that you will also need to specify what version of Node you require using `nvm`, as the default Node version is older than the minimum required for the Firebase Tools CLI.
```
Comment (mlocher) [inline _general/integrations/firebase.md]: We should mention the minimum Node.js version (or at least point to where to find it). I would adapt the sample commands below as well.

File: _general/integrations/firebase.md [hunk 11]
```diff
@@ -0,0 +1,107 @@
...
+### Using Firebase
+
+If your main application uses Firebase as a database or for other services, simply installing the Firebase Tools CLI and setting the required environment variables via your [environment variables]({{ site.baseurl }}{% link _basic/builds-and-configuration/set-environment-variables.md %}) should be all that is required.
+
+However, if you need to pass the Firebase Tools CLI specific commands - such as deploying your static site to Firebase - you can simply add those commands to either your [setup, test or deployment commands]({{ site.baseurl }}{% link _basic/quickstart/getting-started.md %}).
+
+For example, you can add the `firebase deploy` command to a [custom-script deployment pipeline]({{ site.baseurl }}{% link _basic/builds-and-configuration/deployment-pipelines.md %}):
+
+```yaml
+firebase deploy --token "$FIREBASE_TOKEN" --project $PROJECT_NAME
```
Comment (mlocher) [inline _general/integrations/firebase.md]: I would add quotes around `$PROJECT_NAME` as well, and possibly wrap the variable names in braces (see the Pro example above)

File: integrations/index.html [hunk 12]
```diff
@@ -326,6 +326,13 @@ <h4 class="box__title"><a href="{{ site.baseurl }}/tags/digitalocean">Digital Oc
       </div>
 
       <div class="colxs50 colsm50 colmd30 collg20">
+       <div class="box">
```
Comment (mlocher) [inline integrations/index.html]: We should add a helper function to generate those boxes and clean up the integrations page. Could you add that to our trello board?

File: _general/integrations/firebase.md (hunk 1 above)
Comment (ethangj) [reply _general/integrations/firebase.md]: Done

File: integrations/index.html (hunk 12 above)
Comment (ethangj) [reply integrations/index.html]: https://trello.com/c/I8clgSsn

File: _general/integrations/firebase.md (hunk 10 above)
Comment (ethangj) [reply _general/integrations/firebase.md]: I can't find a definitive minimum version. My thought with leaving it without is that if a user has this working locally, they probably have a node version they can peg it to.

**Comment by mlocher** [comment]
@joesiewert could you take another look and approve the PR if it's fine?

**Pull request merged**